### ISSUE TITLE
Not supported evm.bytecode and evm.assembly options in compiler configuration

### ISSUE THREAD

**Issue by smiasojed** [paritytech/revive#39]
Although the tool claims that it does not support `evm.bytecode` and evm.assembly, it still returns it as part of the compilation output.

Used compiler config:
```
{
	"language": "Solidity",
	"settings": {
		"optimizer": {
			"enabled": true,
			"runs": 200
		},
		"outputSelection": {
			"*": {
			"": ["ast"],
			"*": ["abi", "metadata", "devdoc", "userdoc", "storageLayout", "evm.bytecode", "evm.legacyAssembly", "evm.deployedBytecode", "evm.methodIdentifiers", "evm.assembly"]
			}
		}
	}
}

```

It is expected that tool accepts `evm.bytecode` and `evm.assembly`

**Comment by smiasojed** [comment]
`evm.gasEstimates` is also not supported

**Comment by xermicus** [comment]
Stemming from an internal discussion:

- Provide what was requested; however
- Provide default values for incompatible data (if requested)

We substitute any byte-code incompatible fields with empty objects. This won't crash any tools relying on it, they can still request it but will get a zero or empty value. Which they can do nothing with so we should be save. Also this will make remix happy.

**Comment by athei** [comment]
Both of them should be advertised as being available in order to maximize compatibility.

`evm.bytecode` should be returned but empty (we only have init code). `evm.assembly` should contain the PolkaVM dissassembly.

**Comment by xermicus** [comment]
To me it'd also make sense to just provide the blob for both, the deploy code and runtime code? Because it is the same, it seems like we can just provide both for maximum compatibility.

**Comment by smiasojed** [comment]
Remix uses `evm.bytecode` internally, so I think it is better to keep `bytecode` in sync with `deployedBytecode`.

**Comment by athei** [comment]
Because of course they do. They depend on every little detail :(

**Comment by smiasojed** [comment]
> `evm.assembly` should contain the PolkaVM dissassembly.

I think that we need to update polkavm-disassembler to be compatible with evm.legacyAssembly
```
"legacyAssembly": 
				{
							".code": [
								{
									"begin": 199,
									"end": 555,
									"name": "PUSH",
									"source": 0,
									"value": "80"
								}
                                                           ]
				}
```

**Comment by xermicus** [comment]
Legacy assembly is not to confuse with EVM bytecode. Legacy assembly is technically speaking on of our IRs. It can be passed through as-is.

**Comment by xermicus** [comment]
Ah yeah, `assembly_text` is custom,[ it isn't a thing for `solc`](https://docs.soliditylang.org/en/v0.8.28/using-the-compiler.html#output-description).

**Comment by xermicus** [comment]
I reckon either `assembly` and `legacyAssembly` could as well be removed (i.e. replaced with the empty value). We will have maximum compat by keeping it. However I don't see what tools are supposed to do with this.

**Comment by smiasojed** [comment]
"evm.assembly contains assembly_text, so I think it is good to keep it. 
However, Remix displays `evm.legacyAssembly`, which is incorrect.
The options are to either include PolkaVM assembly there or leave it empty.

**Comment by xermicus** [comment]
Ah okay, I see where you are coming from. The problem is we are adding one more IR format. Remix is selling this as the "assembly" code. Which one would assume to be PVM assembly in our case. If we substitute it, we introduce some amount of confusion. But we have to do this anyways for the bytecode, in order to make toolings "just work". For me it's fine to substitute.

**Comment by smiasojed** [comment]
Remix parses this before displaying and expect JSON format. We can not just substitute, because parsing will fail.
In my opinion we have 2 options:
1. empty object
2. polkavm assembly in right format (Requires a change in polkavm-disassembler)

**Comment by xermicus** [comment]
Why would 2. require a change in polkavm-disassembler instead bringing it to the right format directly in revive? Tying the PVM disassembler to this seems a bit stretched 😅

**Comment by xermicus** [comment]
> Stemming from an internal discussion:
> 
>     * Provide what was requested; however
> 
>     * Provide default values for incompatible data (if requested)
> 
> 
> We substitute any byte-code incompatible fields with empty objects. This won't crash any tools relying on it, they can still request it but will get a zero or empty value. Which they can do nothing with so we should be save. Also this will make remix happy.

So if they are incompatible, seems like we hit the second case here.

**Comment by smiasojed** [comment]
> Why would 2. require a change in polkavm-disassembler instead bringing it to the right format directly in revive? Tying the PVM disassembler to this seems a bit stretched 😅
Now looking into it maybe we could indeed parse the polkavm assembly and output something similar to legacyAssembly,
but I am not sure if it make sense.

Legacy assembly is built from objects containing
Name: The opcode (e.g., ADD). 
in polkavm we get `s0 = a0 + a1` so we need to parse the instructions and replace it to ADD
Value: The operand or value pushed to the stack.
In our case we could have registers [s0, a0, a1]
Source: The source code location (related to Solidity source code).
 I think, we do not have it
Begin and End: The bytecode offsets where the instruction starts and ends.
We could get it from disassembler

**Comment by xermicus** [comment]
I vote for just using `assembly_text` directly in remix.

**Comment by xermicus** [comment]
@smiasojed is this issue still relevant?

**Comment by smiasojed** [comment]
> @smiasojed is this issue still relevant?

yes, I need to do still something with `evm.gasEstimates`. I will handle it when I have some time.

**Comment by xermicus** [comment]
@smiasojed I'd like to close this ticket. Is there anything left? I think the gas estimation would justify a dedicated ticket i(f it is required).
